pthread_stop_world.c

```diff
@@ -460,6 +462,49 @@ static int resend_lost_signals(int n_live_threads,
     return n_live_threads;
 }
 
+#ifdef UNITY_RETRY_SIGNALS
+static void suspend_restart_barrier_retry(int n_live_threads, 
+                                          int (*suspend_restart_all)(void))
+{
+#   define TIMEOUT_UNIT 10000
+
+    int i;
+    int acked_threads = 0;
+    struct timespec ts;
+
+    if (clock_gettime(CLOCK_REALTIME, &ts) == -1) {
+        n_live_threads = resend_lost_signals(n_live_threads, suspend_restart_all);
+        suspend_restart_barrier(n_live_threads);
+        return;
+    }
+
+    ts.tv_nsec += TIMEOUT_UNIT * 1000;
+
+    for (i = 0; i < n_live_threads; i++) {
+      while (0 != sem_timedwait(&GC_suspend_ack_sem, &ts)) {
+        /* On Linux, sem_wait is documented to always return zero.      */
+        /* But the documentation appears to be incorrect.               */
+        /* EINTR seems to happen with some versions of gdb.             */
+
+        if (errno == ETIMEDOUT || errno == EINVAL) {
+            // Wait timed out or the timeout period has passed
+            n_live_threads = resend_lost_signals(n_live_threads - acked_threads, suspend_restart_all);
```

Comment:
I don't under the purpose of `acked_threads` being subtracted away and added each failed iteration?

Reply:
There's no failure iteration, it hits only once - if the timeout hits we fallback to the existing resent & barrier and return. `resend_lost_signals` needs to know how many threads it's waiting for (really what value it should expect the semaphore to get to).  In our case that's `n_live_threads - acked_threads`.  `acked_threads` keeps track of how many times we've decremented the semaphore before the timeout.  `resend_lost_signals` can return a different value that it is passed if a thread crashes.

Rather than try and manage the thread count changing, which got a bit complicated, I just did a single failure and fallback to the slow path.  Assuming that if you timeout either you lost a signal or a thread isn't responding quickly, so there's no need to try to optimize.